### DISCUSSION TITLE
Add conversion type hints

### DISCUSSION THREAD

**Discussion by geryogam** [python/typing#1098] [General]
Like the original poster of [this Stack Overflow question](https://stackoverflow.com/q/39505379/2326961), I have some functions with a parameter accepting any arguments that can be converted to `int` (using the `int()` built-in function). This is the type hint that I am currently using (based on [`typeshed`](https://github.com/python/typeshed/blob/master/stdlib/builtins.pyi#L199), and I am not even sure that this list is exhaustive):

```python
from collections.abc import Iterable
from typing import SupportsInt, SupportsIndex, Union

def f(x: Union[SupportsInt, SupportsIndex, str, bytes, bytearray]) -> Iterable:
    return range(int(x))

y = f(1)
y = f(True)
y = f('1')
y = f(b'1')
y = f(bytearray(b'1'))
```

It would be so much more convenient if I did not have to maintain that list.

So what do you think about adding the following type hints to the `typing` module?

- `ConvertibleToInt`, equal to `Union[SupportsInt, SupportsIndex, SupportsTrunc, str, bytes, bytearray]`;
- `ConvertibleToFloat`, equal to `Union[SupportsFloat, SupportsIndex, str, bytes, bytearray]`;
- `ConvertibleToComplex`, equal to `Union[SupportsComplex, SupportsIndex, str, complex]`.

**Comment by srittau** [comment]
It makes sense to me to have these type aliases. For now, the easiest is to add them to the `_typeshed` module, although we plan to move some definitions from `_typeshed` to `typing_extensions`, from where they might migrate to the Python stdlib at some point.

**Reply by geryogam** [reply]
Alright. Is it planned or should I open a pull request?

**Reply by GuillaumeDesforges** [reply]
Same question, can `ConvertibleToInt` be imported and used?

**Comment by srittau** [comment]
`ConvertibleToInt` can be imported from the `_typeshed` module:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from _typeshed import ConvertibleToInt
```

**Reply by GuillaumeDesforges** [reply]
Thanks!
